tortoise/README.md

```diff
@@ -0,0 +1,167 @@
+
+In practice during rerun layer is expected to be finalized after tortoise counted votes from some limited number of layers, this is referred to as verification window in the codebase.  
+
+## Verifying tortoise
+
+Verifying tortoise is meant as a solution for (scaling-issues)[#scaling-issues]of the full tortoise.
+
+Intuitively if there is enough weight from ballots that vote consistently with each other to finalize a layer then tortoise doesn't full vote counting and can use more efficient batched method.
+
+#### Good ballots
+
+Each node maintains a table with local opinions on blocks. For a layer within hdist the opinion is based on the hare output. On layers outside hdist the opinion is based on the decisions from counted votes. Simple example:
+
+| block | layer | opinion |
+| ----- | ----- | ------- |
+| 0x22  | 11    | -1      |
+| 0x33  | 12    | 1       |
+| 0x44  | 13    | 1       |
+
+Important detail that abstain vote is consistent with any local opinion.Otherwise even a single abstain vote will force to switch tortoise into full mode.     
+
+Whenever tortoise receives a ballot it checks if votes from this ballot are consistent with local table:
+
+| block/ballot | layer | opinion | 0xaa | 0xbb |
+| ------------ | ----- | ------- | ---- | ---- |
+| 0x11         | 10    | 1       | 1    | 1    |
+| 0x22         | 11    | -1      | 1    | -1   |
+| 0x33         | 12    | 1       | 1    | 1    |
+| 0x44         | 13    | 1       | -1   | 1    |
+
+In this example only 0xbb ballot is consistent with local opinion. Consistency with a local opinion is a part of the definition of the good ballot, full definition is:
+- ballot has a correct beacon
+- ballot doesn't vote for block before base ballot layer
+- ballot votes are consistent with local opinion 
+- base ballot is good
+
+In the implementation we have a state named "can be good", the role of it's state will be clarified in [modes interaction](#modes-interaction) section. Ballot is marked "can be good" if first 3 condtions are satisfied, but not necessarily the last one.
+
+#### Vote counting in verifying mode
+
+Votes in verifying mode are counted in a batched form, unlike full tortoise. After "goodness" of the ballot is determinted tortoise filters out good ballots and sums up their weight. The result is a counted weight that is consistent with our local opinion. To get an accurate margin that can be compared with global threshold we assume that all uncounted ballots votes against our local opinion.
+
+## Modes interaction
+
+Verifying mode is safer as tortoise doesn't have an imposed limit on the amount of counted votes (or atleast imposed limit can be signicantly higher). Naturally we want to stay in verifying mode as long as tortoise can make progress.
```

Comment:
what is this `imposed limit on the amount of counted votes`?

Reply:
that's about verification window. need to link it somehow